kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/NettyFilterDispatchExecutorTest.java

```diff
@@ -0,0 +1,423 @@
+            CompletionStage<Boolean> chainedAsyncStage = inDispatchThread.thenApplyAsync(o -> dispatchExecutor.isInFilterDispatchThread());
+            completionStage.complete("abc");
+            assertThat(toStandardCompletableFuture(chainedAsyncStage)).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    // sanity check, future is completed as expected when the input future is completed by the dispatcher thread
+    @Test
+    void completeOnFilterDispatchThreadWhenInDispatchThread() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Boolean> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage)
+                    .thenApply(o -> dispatchExecutor.isInFilterDispatchThread());
+            dispatchExecutor.execute(() -> completionStage.complete("arbitrary"));
+            assertThat(toStandardCompletableFuture(inDispatchThread)).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void completeOnFilterDispatchThreadWhenCompletedByAnotherEventLoop() {
+        EventLoop eventLoop = new DefaultEventLoop();
+        EventLoop anotherLoop = new DefaultEventLoop();
+        try {
+            NettyFilterDispatchExecutor dispatchExecutor = new NettyFilterDispatchExecutor(eventLoop);
+            CompletableFuture<Object> completionStage = new CompletableFuture<>();
+            CompletionStage<Boolean> inDispatchThread = dispatchExecutor.completeOnFilterDispatchThread(completionStage)
+                    .thenApply(o -> dispatchExecutor.isInFilterDispatchThread());
+            anotherLoop.execute(() -> completionStage.complete("arbitrary"));
+            assertThat(toStandardCompletableFuture(inDispatchThread)).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+            // double-check we are on the expected eventloop
+            CompletableFuture<Object> completionStage2 = new CompletableFuture<>();
+            CompletionStage<Boolean> inExpectedEventLoop = dispatchExecutor.completeOnFilterDispatchThread(completionStage2)
+                    .thenApply(o -> eventLoop.inEventLoop() && !anotherLoop.inEventLoop());
+            anotherLoop.execute(() -> completionStage2.complete("arbitrary"));
+            assertThat(toStandardCompletableFuture(inExpectedEventLoop)).succeedsWithin(5, TimeUnit.SECONDS).isEqualTo(true);
+        }
+        finally {
+            eventLoop.shutdownGracefully();
+            anotherLoop.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void scheduleCallableDelegated() {
```

Comment:
I just want to voice my thoughts.   See the mock gave me pause for thought. 

You are asserting implementation here rather than behaviour.   The fact that it uses a delegation pattern is really implementational detail.  Really, the important aspect is the callable is run / can be cancelled - not how the implementation actually works underneath.     A puritan approach would have tests written exercising those cases..  However, I think in this situation you've done the pragmatic thing and sensibly avoid mimicking tests that will exist in the JDK.